src/main/java/com/minecolonies/coremod/colony/workorders/WorkOrderBuildDecoration.java

```diff
@@ -19,29 +51,309 @@ public WorkOrderBuildDecoration()
+        }
+        else
+        {
+            compound.setString(TAG_SCHEMATIC_NAME, structureName);
+        }
+        compound.setInteger(TAG_BUILDING_ROTATION, buildingRotation);
+        compound.setBoolean(TAG_IS_REQUESTED, requested);
+        compound.setBoolean(TAG_IS_MIRRORED, isMirrored);
+    }
+
+    /**
+     * Attempt to fulfill the Work Order.
+     * Override this with an implementation for the Work Order to find a Citizen to perform the job
+     * <p>
+     * finds the first suitable builder for this job.
+     *
+     * @param colony The colony that owns the Work Order.
+     */
+    @Override
+    public void attemptToFulfill(@NotNull final Colony colony)
+    {
+        boolean sendMessage = true;
+        boolean hasBuilder = false;
+
+        for (@NotNull final CitizenData citizen : colony.getCitizens().values())
+        {
+            final JobBuilder job = citizen.getJob(JobBuilder.class);
+
+            if (job == null || citizen.getWorkBuilding() == null)
+            {
+                continue;
+            }
+
+            hasBuilder = true;
+
+            // don't send a message if we have a valid worker that is busy.
+            if (canBuild(citizen))
+            {
+                sendMessage = false;
+            }
+
+            if (!job.hasWorkOrder() && canBuild(citizen))
+            {
+                job.setWorkOrder(this);
+                this.setClaimedBy(citizen);
+                return;
+            }
+        }
+
+        sendBuilderMessage(colony, hasBuilder, sendMessage);
+    }
+
+    /**
+     * send a message from the builder.
+     *
+     * Suppressing Sonar Rule squid:S1172
+     * This rule does "Unused method parameters should be removed"
+     * But in this case extending class may need to use the sendMessage parameter
+     * @param colony which the work order belong to
+     * @param hasBuilder true if we have a builder for this work order
+     * @param sendMessage true if we need to send the message
+     */
+    @SuppressWarnings("squid:S1172")
+    protected void sendBuilderMessage(@NotNull final Colony colony, final boolean hasBuilder, final boolean sendMessage)
+    {
+        if (hasSentMessageForThisWorkOrder)
+        {
+            return;
+        }
+
+        if (!hasBuilder)
+        {
+            hasSentMessageForThisWorkOrder = true;
+            LanguageHandler.sendPlayersMessage(colony.getMessageEntityPlayers(),
+              "entity.builder.messageNoBuilder");
+        }
+    }
+
+    /**
+     * Checks if a builder may accept this workOrder.
+     *
+     * Suppressing Sonar Rule squid:S1172
+     * This rule does "Unused method parameters should be removed"
+     * But in this case extending class may need to use the citizen parameter
+     * @param citizen which could build it or not
+     * @return true if he is able to.
+     */
+    @SuppressWarnings("squid:S1172")
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Suppressing the 'squid:S1172' warning is not allowed [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1309)